keps/sig-storage/20190120-ExecutionHook-apichange.md

```diff
@@ -0,0 +1,146 @@
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+      * [Workarounds](#workarounds)
+      * [Alternatives](#alternatives)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+
+## Summary
+
+Volume snapshot support was introduced in Kubernetes v1.12 as an alpha feature. 
+In the alpha implementation of snapshots for Kubernetes, there is no snapshot consistency
+guarantees beyond any guarantees provided by storage system (e.g. crash consistency).
+
+This proposal is aimed to address that limitation by providing an `ExecutionHook`
+in the `Container` struct. The snapshot controller will look up this hook before
+taking a snapshot and execute it accordingly.
+
+## Motivation
+
+The volume snapshot feature allows creating/deleting volume snapshots, and the
+ability to create new volumes from a snapshot natively using the Kubernetes API.
+
+However, application consistency is not guaranteed. An user has to figure out how
+to quiece an application before taking a snapshot and unquiece it after taking
+the snapshot.
+
+So we want to introduce an `ExecutionHook` to facilitate the quiesce and unquiesce
+actions when taking a snapshot. There is an existing lifecycle hook in the `Container`
+struct. The lifecycle hook is called immediately after a container is created or
+immediately before a container is terminated. The proposed execution hook is not
+tied to the start or termination time of the container. Instead it is called before
+or after taking a snapshot while the container is still running.
+
+### Goals
+
+`ExecutionHook` is introduced to define actions that can be taken on a container.
+Specifically it can be used to perform a quiese operation before a volume snapshot
+is taken and then perform an unquiesce operation after a volume snapshot is
+taken to resume the application.
+
+## Non-Goals
+
+This proposal does not provide exact command included in the `ExecutionHook`
+because every application has a different requirement.
+
+## Proposal
+
+An `ExecutionHook` is defined in the following.
+
+```
+// ExecutionHook defines a specific action that should be taken with retries and timeouts
+type ExecutionHook struct {
+        // Command to execute for a particular trigger
+        Handler `json:",inline" protobuf:"bytes,1,opt,name=handler"`
+        // How long the controller should try/retry to execute the hook before giving up
+        RetryTimeOutSeconds int64 `json:"retryTimeOutSeconds,omitempty" protobuf:"varint,2,opt,name=retryTimeOutSeconds"`
+        // Number of retries
+        NumRetries int64 `json:"numRetries,omitempty" protobuf:"varint,3,opt,name=numRetries"`
+}
+```
+
+A `QuiesceUnquiesceHook` struct contains a `Quiesce` and `Unquiesce` `ExecutionHook`.
+
+```
+// QuiesceUnquiesceHook includes a Quiesce action and an Unquiesce action
+type QuiesceUnquiesceHook struct {
```

Comment:
First the name - this is very hard to spell.  In fact this doc gets it wrong as often as not.  Can we adopt simpler naming?  Pause/Resume or something that is meaningful but less difficult?

Reply:
As a point to discuss - why do we want this as a specific hook?  Assuming we do this at all, would we rather a general "notification" or "imperative" API or N specific hooks?

An advantage of being specific is that we can make specific sub-resources that we can ACL specifically.

An advantage of being generic is that we don't have to keep solving this in almost-identical ways.

@liggitt @lavalamp @deads2k

Reply:
Please see the updated spec. Introduced a generic ExecutionHook but also added a type to indicate it is for "Freeze"/"Thaw".  Support for new types can be added later.